feat: virtual doctype for frappe recorder

GursheenK opened the pull request:
### Made a Virtual Doctype for **Frappe Recorder**.
<br>
The list view for the Recorder looks roughly like this with the backend functionality of Start / Stop / Clear remaining the same - 

<img width="1290" alt="Screenshot 2023-08-05 at 5 52 05 PM" src="https://github.com/frappe/frappe/assets/40693548/23927897-0277-47bb-b3cd-16fff768336e">

<br>
<br>

Made a Child Table for storing **SQL Queries** for a particular request in the form view for each request.

<img width="963" alt="Screenshot 2023-08-03 at 12 14 46 PM" src="https://github.com/frappe/frappe/assets/40693548/2de99ea7-5eae-49da-a070-951abf083b25">

<br>

<img width="1198" alt="Screenshot 2023-08-05 at 5 53 02 PM" src="https://github.com/frappe/frappe/assets/40693548/4cdf53a7-0683-49d4-b0f2-abe3efc64d0d">

<br>

**TODO**

- [x] Add Stack Trace in Child Table SQL Queries.
- [x] Add Import / Export functionality for requests.
- [x] Add appropriate filters.

Need a clearer picture for -

1. Virtual Child Table changes made.
2. Whether Stack Trace should be a child table inside the SQL Queries child table itself or just a field.
3. Any changes required for printing JSON fields like Form Dict and Request Headers.

`no-docs`

GursheenK committed:
feat: allow virtual child tables

GursheenK committed:
feat: add child table for recorder requests

GursheenK committed:
feat: add child table for sql queries

ankush commented:
Notes for reviewer.

Main things to reivew:
- [ ] virtual child table API design


Other things:
- [x] starting/stopping/clearing recorder
- [x] import/export
- [x] Sorting on list view should be by duration/queries by default (and should support most sensible sort keys)
- [x] child table sorting by idx, duration, copies etc (?) (This is deferred, will be generic feature someday later)
- [x] Filtering on list view
- [x] Remove old recorder and change route/names
- [x] Realtime listview changes

ankush commented on frappe/core/doctype/recorder_request/recorder_request.py:
```diff
@@ -0,0 +1,211 @@
+from frappe.database.database import is_query_type
+from frappe.model.document import Document
+
+RECORDER_INTERCEPT_FLAG = "recorder-intercept"
+RECORDER_REQUEST_SPARSE_HASH = "recorder-requests-sparse"
+RECORDER_REQUEST_HASH = "recorder-requests"
+TRACEBACK_PATH_PATTERN = re.compile(".*/apps/")
```
We can leave `recorder.py` code as is and just import what is necessary IMO. Backend code won't require any changes for this PR. No need copy things here.

ankush commented:
@GursheenK 

> Whether Stack Trace should be a child table inside the SQL Queries child table itself or just a field.

It can just be a code field inside the row. We don't support nested child table anyway. Similar thing has to be done for "Explain" output too. If the code field isn't readable we can add an HTML field with some small formatting apt for data. 

> Any changes required for printing JSON fields like Form Dict and Request Headers.

Just formatting them should be enough. Pass it through `frappe.as_json` and it will be readable enough for developers.

ankush commented:
@Alchez IIRC you had tried to implement virtual child tables. Can you share your exp / possible design. 


IMO parent doctype controllers should be responsible for fetching and initializing virtual child tables. Direct access to virtual child docs won't be supported. Please share if you have any better proposals for design.

In terms of CRUD:
- Create: parent can insert child document if it's not found (indicated by doc.is_new()) 
- Read: parent can fetch all child and init them
- Update: db_update can update all virtual child tables too. 
- Delete: This is tricky. Framework does diffing and removes old rows from DB. Something similar will have to be written in EACH virtual doctype controller. 


cc: @surajshetty3416

ankush commented on frappe/core/doctype/recorder_request/recorder_request_list.js:
```diff
@@ -0,0 +1,47 @@
+				}
+			});
+
+		listview.page.add_button(__("Start"), () => {
```
Start/stop buttons should be primary and only one (relevant one) should be shown at a time.

Alchez commented:
@ankush 

> @Alchez IIRC you had tried to implement virtual child tables. Can you share your exp / possible design.

Yeah, we have a read-only implementation of a virtual child table under an existing doctype. Since we didn't require all the CRUD operations, we were able to make it work by doing the following:
- Setup the `get_list` method in the virtual child table controller.
- Override the parent doctype's controller and use the `__setup__` hook defined in [`document.py`](https://github.com/frappe/frappe/blob/fa73d5ea03273fc88e23fb0cd80df3f6bc9f8836/frappe/model/document.py#L199) to call the `get_list` function and set the child table dynamically.

Another viable option would have been to override the `load_from_db` method in the parent, but since that would be a bigger piece to re-write, and `__setup__` is called at the end of `load_from_db`, we decided to use that one.

> IMO parent doctype controllers should be responsible for fetching and initializing virtual child tables. Direct access to virtual child docs won't be supported.

That makes sense to me, considering that tables are dependent on the parent doctype for context. But we should probably cover some generic use-cases for this before locking that design.

> In terms of CRUD:
> - Create: parent can insert child document if it's not found (indicated by doc.is_new())
> - Read: parent can fetch all child and init them
> - Update: db_update can update all virtual child tables too.
> - Delete: This is tricky. Framework does diffing and removes old rows from DB. Something similar will have to be written in EACH virtual doctype controller.

Those look like good guidelines for a virtual child doctype. The parent document would most likely always be responsible for the table.

GursheenK committed:
fix: remove redundant fns from recorder request

GursheenK committed:
feat: add html fields for stack trace and sql explain

GursheenK committed:
feat: add js for rendering html fields

GursheenK committed:
fix: indicators and primary btn in listview

GursheenK committed:
fix: handle file uploader obj while importing

GursheenK committed:
fix: show idx for stack trace

GursheenK committed:
fix: filter and sort list view

GursheenK committed:
test: allow parent and child docs of same type

GursheenK committed:
test: crud with virtual child table

ankush committed:
fix(DX): enable realtime refresh on recorder listview

ankush committed:
fix: fast primary button switching

ankush commented on frappe/core/doctype/recorder_request/recorder_request.json:
```diff
@@ -0,0 +1,126 @@
+   "fieldtype": "Select",
+   "in_standard_filter": 1,
+   "label": "Method",
+   "options": "GET\nPOST"
```
There are more :grimacing:

ankush committed:
fix: Add all HTTP methods

ankush committed:
refactor: code cleanup and fix filters

- rename recorder.get
- move unused methods below
- fix "like" filter
- move filters out to separate function
- revert doctype change

ankush commented on frappe/core/doctype/recorder_request/recorder_request.py:
```diff
@@ -0,0 +1,100 @@
+			for field in filters:
+				operator = "in" if filters[field][0] == "like" else filters[field][0]
+				if not compare(request[field], operator, filters[field][1]):
+					filter_flag = 0
```
You can avoid flags like these and directly append thing to list when condition matches. 


(pushing changes)

ankush committed:
fix: include creation/modified timestamp

ankush committed:
chore: hide comment count

ankush committed:
fix: if request is deleted then 404

ankush committed:
refactor: delete old recorder and rename new one

ankush committed:
fix: missing order by clause and tests

ankush commented:
@GursheenK tests are missing for virtual doctype. I added one here https://github.com/frappe/frappe/pull/21908/commits/8ffd363ccf2147164b4e6b18531a41177841b102, can you add some more:

- [x] Filtering (path, query count) and ordering correctness
- [x] Serialization correctness

Examples: https://github.com/frappe/frappe/blob/8ffd363ccf2147164b4e6b18531a41177841b102/frappe/core/doctype/rq_job/test_rq_job.py#L58-L81

ankush committed:
refactor: flatten file argument in function call

ankush committed:
Merge branch 'develop' into virtual-doc-for-frappe-recorder

ankush committed:
fix: delete imported recorder file after import

GursheenK committed:
test: recorder filter sort and serialization

surajshetty3416 commented on frappe/core/doctype/recorder/recorder.js:
```diff
@@ -0,0 +1,47 @@
+frappe.ui.form.on("Recorder Query", "form_render", function (frm, cdt, cdn) {
+	let row = frappe.get_doc(cdt, cdn);
+	let stack = JSON.parse(row.stack);
+	render_html_field(stack, "stack_html", "Stack Trace");
```
Please make sure that user facing text is translatable.
```suggestion
	render_html_field(stack, "stack_html",  __("Stack Trace"));
```

surajshetty3416 commented on frappe/core/doctype/recorder/recorder.js:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	function create_html_table(table_content, html) {
+		html +=
+			"<div class='control-value like-disabled-input for-description' style='overflow:auto; padding:0px'>";
+		html += "<table class='table table-striped' style='margin:0px'><thead><tr>";
+		Object.keys(table_content[0]).forEach((key) => {
+			html += "<th>" + key + "</th>";
+		});
+		html += "</tr></thead><tbody>";
+
+		for (let row in table_content) {
+			html += "<tr>";
+			Object.values(table_content[row]).forEach((value) => {
+				html += "<td>" + value + "</td>";
+			});
+			html += "</tr>";
+		}
+		html += "</tbody></table></div>";
+		return html;
```
This code block has poor readability, better to use [template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) while constructing such multiline html block.

cc: @GursheenK @ankush


```suggestion
html += `
	<div class='control-value like-disabled-input for-description'
		style='overflow:auto; padding:0px'>
		<table class='table table-striped' style='margin:0px'>
			<thead>
				<tr>
					${
						Object.keys(table_content[0])
							.map(key => `<th>${key}<th>`)
							.join("")
						}
				</tr>
			</thead>
			<tbody>
				${
					table_content.map(content => {
						return `
							<tr>
								${
									Object.values(content)
										.map(key => `<td>${key}<td>`)
										.join("")
								}
							</tr>
						`
					}).join("")
				}
			</tbody>
		</table>
	</div>
`;

return html
```

barredterra commented on frappe/patches.txt:
```diff
@@ -193,6 +193,7 @@ frappe.patches.v14_0.delete_payment_gateways
 frappe.patches.v15_0.remove_event_streaming
 frappe.patches.v15_0.copy_disable_prepared_report_to_prepared_report
 execute:frappe.reload_doc("desk", "doctype", "Form Tour")
+execute:frappe.delete_doc('Page', 'recorder', ignore_missing=True, force=True)
```
@GursheenK this seems to cause the following error message for me:

```
frappe.exceptions.ValidationError: Deletion of this document is only permitted in developer mode.
```
